dist/bin/scala

```diff
@@ -141,8 +150,20 @@ if [ $execute_script == true ]; then
     scala_exit_status=$?
   else
     [[ $save_compiled == true ]] && rm -f $target_jar
-    residual_args+=($setScriptName)
-    eval "\"$PROG_HOME/bin/scalac\" ${cp_arg-} ${java_options[@]} ${residual_args[@]} -script $target_script ${script_args[@]}"
+    set -- ${cp_arg-} ${java_options[@]} ${residual_args[@]} -script "$target_script" ${script_args[@]}
+    PROG_MAIN=$ScriptingMain
+    prepScalacCommandLine "$@"
```

Comment:
I'm wondering what arguments are expected to be parsed here? If it's only 1 or 2, can we do it in `bin/scala` instead?

Reply:
I considered trying to do a less extensive change, but ran into problems. 

One concern is has to do with alternate ways for launching a script.   The following two command lines are currently supported valid ways to launch a script:

```bash
scala hello.sc
scalac -script hello.sc
```
These and other similar issues are tested in `project/scripts/winCmdTests`, and possibly elsewhere.

My assumption is that we want a functionally neutral refactoring, so scripts should use the same default values, e.g.:
```bash
default_java_opts="-Xmx768m -Xms768m"
```
They should leverage a single toolchain classpath, and should have access to and respond the same way to scalac options such as `-Oshort`, for example.  So most of what happens in `bin/scalac` is needed to produce a functionally equivalent java command line.

Anyway, that's what led me to the current implementation, which can no doubt be improved in various ways.

Reply:
> scalac -script hello.sc

I am not sure if Scala 2 supports the above. If not, I think it does not have to be supported.

> My assumption is that we want a functionally neutral refactoring, so scripts should use the same default values, for example default_java_opts="-Xmx768m -Xms768m"

This can be either duplicated or moved to `bin/common` -- maybe duplicating it is easier for maintenance.

> They should leverage a single toolchain classpath, and should have access to and respond the same way to scalac options such as -Oshort, for example. So most of what happens in bin/scalac is needed to produce a functionally equivalent java command line.

Given that `bin/scala` already parses many options that `bin/scalac` parses, it does not matter to parse 1-2 more options.

Reply:
> Given that `bin/scala` already parses many options that `bin/scalac` parses, it does not matter to parse 1-2 more options.

Okay, I see what you have in mind.  I'll will review it again.

> I am not sure if Scala 2 supports the above. If not, I think it does not have to be supported.

I don't think it's supported by scala 2, although both ways of launching the script must provide `-script <scriptname>`  on the generated java command line, because  [compiler/src/dotty/tools/scripting/Main.scala](https://github.com/lampepfl/dotty/blob/2290dbca91cbf98aec9f039e011aa43dc39281c7/compiler/src/dotty/tools/scripting/Main.scala)  line 12 uses it to distinguish jvm parameters from script filename and args.

Reply:
@liufengyun
Is your idea that we move the command line parsing from `bin/common` back to `bin/scalac` and selectively duplicate some of the needed parsing in `bin/scala` ?